controller/root.go

```diff
@@ -0,0 +1,178 @@
+func NewRootController(service *goa.Service) *RootController {
+	return &RootController{
+		Controller:  service.NewController("RootController"),
+		FileHandler: workingFileFetcher{},
+	}
+}
+
+// List runs the list action.
+func (c *RootController) List(ctx *app.ListRootContext) error {
+	roots, err := getRoot(ctx.Request, c.FileHandler)
+	if err != nil || roots == nil {
+		return jsonapi.JSONErrorResponse(
+			ctx, err)
+	}
+
+	res := &app.RootSingle{}
+	res.Data = roots
+	return ctx.OK(res)
+}
+
+// Get a list of all endpoints formatted to json api format.
+func getRoot(request *http.Request, fileHandler asseter) (*app.Root, error) {
+	// Get an unmarshal swagger specification
+	result, err := getUnmarshalledSwagger(fileHandler)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get and iterate over paths from swagger specification.
+	swaggerPaths, err := getSwaggerFieldAsMap("paths", result)
+	if err != nil {
+		return nil, err
+	}
+
+	namedPaths := make(map[string]interface{})
+	for path, swaggerPath := range swaggerPaths {
+
+		// Currently not supporting endpoints that contain parameters.
+		if !strings.Contains(path, "{") {
+
+			// Use the segments in the path to construct a meaningful name to use for the path.
+			key := strings.Replace(path, "/", "_", -1)
+			key = strings.Replace(key, "_", "", 1)
+
+			pathsObj, err := swaggerPath.(map[string]interface{})
+			if !err {
+				return nil, errorLogAndReturn("Unable to assert correct format for swagger specifiation")
+			}
+
+			// Get the x-tag value. If the tag exists, use it as path name.
+			xtag, _ := getSwaggerFieldAsString("x-tag", pathsObj, false)
+			if len(xtag) > 0 {
+				key = xtag
+			}
+
+			// Set the related field and link objects for each name.
+			namedPaths[key] = map[string]interface{}{
+				"links": map[string]string{
+					"related": path,
+				},
+			}
+		}
+	}
+
+	id := uuid.NewV4()
+	basePath, err := getSwaggerFieldAsString("basePath", result, true)
+	if err != nil {
+		return nil, err
+	}
+
+	basePath = rest.AbsoluteURL(request, basePath)
+	links := &app.GenericLinksForRoot{
+		Self: &basePath,
+	}
+	return &app.Root{Relationships: namedPaths, ID: &id, Links: links}, nil
+}
+
+// Gets the swagger specification binary and attempts to unmarshal it.
+// Returns the unmarshed specification, or error. Error is returned if
+// the specification could not be found, or the specification was not
+// able to be unmarshalled.
+func getUnmarshalledSwagger(fileHandler asseter) (map[string]interface{}, error) {
+	swaggerJSON, err := fileHandler.Asset("swagger.json")
+	if err != nil {
+		log.Error(nil, map[string]interface{}{
+			"file": "swagger.json",
+		}, "The file with id 'swagger.json' not found")
+		return nil, errors.NewNotFoundError("file", "swagger.json")
+	}
+
+	var result map[string]interface{}
+	err = json.Unmarshal([]byte(swaggerJSON), &result)
+	if err != nil {
+		return nil, errorLogAndReturn("Unable to unmarshal json swagger specification")
```

Comment:
You don't log nor return the error which is not good. Please make sure to also wrap an error when it occurs. Search for `errs.Wrapf(` in our code to see how it is done.